Kubernetes support for DVMA

seizadi opened the pull request:
There was not support for running DVMA on containerized environments in the project added support for building a container and chart to run it on Kubernetes. This is using my docker registry soheileizadi but I assume you will change it to a community registry and crate Github action to build it. It is also working with older 1.9 version of the project not the latest.
You can review build and deployment guide in [README](https://github.com/seizadi/DVWA/blob/master/deploy/README.md)

digininja commented:
I know absolutely nothing about Kubernetes so would not be able to get any of this working or support any of it.

If you want to submit a fully working feature along with all your contact details for support then I'll consider adding it.

seizadi commented:
HI Robin,
Thanks for the project, I have found it very useful. What I checked in does work, I'm willing to get on a zoom call and step you through it if you like.

digininja commented on deploy/conf/setup_dvwa.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+chmod -R 777 /app/hackable/uploads /app/external/phpids/0.6/lib/IDS/tmp/phpids_log.txt && \
```
the log file shouldn't be executable

digininja commented:
I've just had a look and you'll need to update it to use GitHub Master. We don't release versions and only support the Git Head.

That means PHP 7.3 and MariaDB. I can't see anything about installing the database server do don't know if you are doing MySQL or MariaDB.

seizadi commented:
This was to work with 1.9 tag version of DVWA, so it is using PHP5 and MySQL. I don't have any experience with the Git Head version of DVWA, let me work on that and I will submit a PR once I get that working.

digininja commented:
the version 1.9 is from 2015, there is no way we can support that.

seizadi commented:
You have tags/releases in the project, you can tag the image and version chart so that they work at a particular time with the tags and meanwhile the latest image/chart would work with Git Head/master. I will work on getting this to work with Git Head.

digininja commented:
We have a few releases but they are only when someone remembers to do them or someone asks for a release, they are rarely fully up to date with the current master branch.

seizadi commented:
I see a container runtime, the image is a couple years old, but I don't see a Dockerfile in this project for it, trying to figure out if that is a good start for a newer image: https://github.com/digininja/DVWA#docker-container

digininja commented:
No idea, as with kubernetes, someone else submitted that and said they
would handle support. If I needs updating, let me know what needs doing.

On Fri, 18 Sep 2020, 18:37 Soheil Eizadi, <notifications@github.com> wrote:

> I see a container runtime, the image is a couple years old, but I don't
> see a Dockerfile in this project for it, trying to figure out if that is a
> good start for a newer image:
> https://github.com/digininja/DVWA#docker-container
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/digininja/DVWA/pull/384#issuecomment-694996205>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4SWOOTU57W7DXBM6AMR3SGOLE5ANCNFSM4RQSXE2A>
> .
>

seizadi commented:
If you are happy with that image I will use it for Kubernetes. I checked and it is using PHP7
root@277cbd9fe403:/var/www/html# php -v
PHP 7.0.30-0+deb9u1 (cli) (built: Jun 14 2018 13:50:25) ( NTS )

I will remove the Dockerfile from PR, and reference this container.

digininja commented:
Seeing as that install uses its own copy of DVWA from its own repo, that will be very out of date by now.

https://github.com/opsxcq/docker-vulnerable-dvwa

As I said, I don't support it, but it looks like the @opsxcq doesn't support it any longer either.

The best thing you can do is to just do a git pull and then you are guaranteed to get the latest version.

seizadi commented:
If have docker, you can run the 'make run' target and have it run on your system. This is what the front page looks like:

```bash
Database Setup 
Click on the 'Create / Reset Database' button below to create or reset your database.
If you get an error make sure you have the correct user credentials in: /app/config/config.inc.php

If the database already exists, it will be cleared and the data will be reset.
You can also use this to reset the administrator credentials ("admin // password") at any stage.


Setup Check
Operating system: *nix
Backend database: MySQL
PHP version: 7.2.24-0ubuntu0.18.04.6

Web Server SERVER_NAME: localhost

PHP function display_errors: Disabled
PHP function safe_mode: Disabled
PHP function allow_url_include: Enabled
PHP function allow_url_fopen: Enabled
PHP function magic_quotes_gpc: Disabled
PHP module gd: Installed
PHP module mysql: Installed
PHP module pdo_mysql: Installed

MySQL username: dvwa
MySQL password: ******
MySQL database: dvwa
MySQL host: 127.0.0.1

reCAPTCHA key: Missing

[User: www-data] Writable folder /app/hackable/uploads/: Yes
[User: www-data] Writable file /app/external/phpids/0.6/lib/IDS/tmp/phpids_log.txt: Yes


[User: www-data] Writable folder /app/config: Yes
Status in red, indicate there will be an issue when trying to complete some modules.

If you see disabled on either allow_url_fopen or allow_url_include, set the following in your php.ini file and restart Apache.
allow_url_fopen = On
allow_url_include = On
These are only required for the file inclusion labs so unless you want to play with those, you can ignore them.
```

digininja commented:
What version of the code is it using?

seizadi commented:
It pull master.zip

digininja commented on deploy/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+# Install packages
+ENV DEBIAN_FRONTEND noninteractive
+RUN apt-get update && \
+		apt-get -y install git wget unzip supervisor apache2 mysql-server php php-mysqli php-gd libapache2-mod-php
```
Can we have MariaDB rather than MySQL, there are loads of little issues with MySQL that make it a pain to support

digininja commented on deploy/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+# Preparation
+RUN \
+#  rm -fr /app/* && \
+  wget https://github.com/ethicalhack3r/DVWA/archive/master.zip && \
```
it is on digininja not ethicalhack3r now

digininja commented on deploy/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+  unzip /master.zip && \
+  cp -r /DVWA-master/* /app && \
+  rm -rf /DVWA-master && \
+  echo 'session.save_path = "/tmp"' >> /etc/php/7.2/apache2/php.ini && \
```
is it always going to be PHP 7.2 that is installed? If it gets updated to 7.3 then this will break.

I don't know if Ubuntu would change a major version in an LTS release or not

fldu replied:
You can put the php version as an ARG at the beginning of the Dockerfile, but you'll have to precise the version of the package in the 'apt install' line

digininja commented on deploy/conf/apache_default.conf:
```diff
@@ -0,0 +1,38 @@
+	#
+	# Set HTTPS environment variable if we came in over secure
+	#  channel.
+	SetEnvIf x-forwarded-proto https HTTPS=on
```
Can drop this bit as you aren't setting up a HTTPS server

digininja commented on deploy/conf/apache_default.conf:
```diff
@@ -0,0 +1,38 @@
+                allow from all
+        </Directory>
+
+        ScriptAlias /cgi-bin/ /usr/lib/cgi-bin/
```
cgi-bin isn't used so isn't needed

digininja commented on deploy/conf/create_mysql_users.sh:
```diff
@@ -0,0 +1,63 @@
+mysql -uroot -e "CREATE USER 'admin'@'%' IDENTIFIED BY '$PASS'"
+mysql -uroot -e "GRANT ALL PRIVILEGES ON *.* TO 'admin'@'%' WITH GRANT OPTION"
+
+mysql -uroot -e " GRANT ALL PRIVILEGES ON phpmyadmin.* TO  'pma'@'localhost' IDENTIFIED BY ''"
```
phpmyadmin isn't being installed is it?

digininja commented on deploy/conf/run.sh:
```diff
@@ -0,0 +1,112 @@
+    rm -f /var/www/html && ln -s "/app/${APACHE_ROOT}" /var/www/html
+fi
+
+echo "Editing phpmyadmin config"
```
is phpmyadmin being installed?

digininja commented on README.md:
```diff
@@ -139,10 +139,10 @@ Login URL: http://127.0.0.1/dvwa/login.php
+`docker run --rm -it -p 80:80 soheileizadi/dvwa`
 
-Please ensure you are using aufs due to previous MySQL issues. Run `docker info` to check your storage driver. If it isn't aufs, please change it as such. There are guides for each operating system on how to do that, but they're quite different so we won't cover that here.
+The [Dockerfile](deploy/Dockerfile) in the project is available if you like to customize your container.
```
Missing word

"if you WOULD like to"

digininja commented:
Can you talk to @peter279k about a local docker setup. I'd rather use the other repo but as it isn't being supported I guess it is better to bring it in-house. Your install looks more robust.

fldu commented on deploy/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+FROM ubuntu:bionic
+MAINTAINER soheil@gmail.com
+
+ENV CREATE_MYSQL_BASIC_USER_AND_DB true
+ENV MYSQL_USER_NAME dvwa
+ENV MYSQL_USER_DB dvwa
+ENV MYSQL_USER_PASS p@ssw0rd
```
Might be interesting to move this into environment file

seizadi replied:
Is the intent to allow users to change set their own password using environment variables? I put this here so it was obvious that additional user and database was being created for DVMA, if someone changes the php.config they know to change it here also, I could have hard-coded it in the create_mysql_users.sh. but instead made them configured at this top level.

seizadi commented:
Added additional ARGs. The PHP version change is actually the side-effect of picking the latest version released with each distribution. In Ubuntu 16.04 it was 7.0 and with Ubuntu Bionic (18.04) it is 7.2, I assume it will drift if we upgrade the distribution.

The database ARG I added but it will not work for MariaDB, I set it and it failed and I don't think it is a good use of my time to get two databases working with a single image, here are the error messages you get when you set it:
```bash
Editing MySQL config
sed: can't read /etc/mysql/mysql.conf.d/mysqld.cnf: No such file or directory
sed: can't read /etc/mysql/mysql.conf.d/mysqld.cnf: No such file or directory
=> An empty or uninitialized MySQL volume is detected in /var/lib/mysql
=> Installing MySQL ...
2020-09-22  2:58:21 140640638819456 [Note] mysqld (mysqld 10.1.44-MariaDB-0ubuntu0.18.04.1) starting as process 22 ...
2020-09-22  2:58:21 140640638819456 [ERROR] mysqld: Can't create/write to file '/var/lib/mysql/aria_log_control' (Errcode: 13 "Permission denied")
2020-09-22  2:58:21 140640638819456 [ERROR] mysqld: Got error 'Can't create file' when trying to use aria control file '/var/lib/mysql/aria_log_control'
2020-09-22  2:58:21 140640638819456 [ERROR] Plugin 'Aria' init function returned error.
2020-09-22  2:58:21 140640638819456 [ERROR] Plugin 'Aria' registration as a STORAGE ENGINE failed.
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: innodb_empty_free_list_algorithm has been changed to legacy because of small buffer pool size. In order to use backoff, increase buffer pool at least up to 20MB.

2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Using mutexes to ref count buffer pool pages
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: The InnoDB memory heap is disabled
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Compressed tables use zlib 1.2.11
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Using Linux native AIO
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Using SSE crc32 instructions
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Initializing buffer pool, size = 128.0M
2020-09-22  2:58:21 140640638819456 [Note] InnoDB: Completed initialization of buffer pool
2020-09-22 02:58:21 7fe973515c80  InnoDB: Operating system error number 13 in a file operation.
InnoDB: The error means mysqld does not have the access rights to
InnoDB: the directory.
2020-09-22  2:58:21 140640638819456 [ERROR] InnoDB: os_file_get_status() failed on './ibdata1'. Can't determine file permissions
2020-09-22  2:58:21 140640638819456 [ERROR] InnoDB: The system tablespace must be writable!
2020-09-22  2:58:21 140640638819456 [ERROR] Plugin 'InnoDB' init function returned error.
2020-09-22  2:58:21 140640638819456 [ERROR] Plugin 'InnoDB' registration as a STORAGE ENGINE failed.
2020-09-22  2:58:21 140640638819456 [Note] Plugin 'FEEDBACK' is disabled.
2020-09-22  2:58:21 140640638819456 [ERROR] Could not open mysql.plugin table. Some plugins may be not loaded
2020-09-22  2:58:21 140640638819456 [ERROR] mysqld: unknown option '--initialize-insecure'
2020-09-22  2:58:21 140640638819456 [ERROR] Aborting

=> Done!
```

seizadi commented:
The red warning about ReCAPTCHA was bugging me so I fixed it so if you have one you can push it in environment variables and it will set it and show the public key. Updated the README to show how to set this up with docker.

seizadi commented:
I think I've incorporated all your change requests.

digininja commented:
Is this ready to be merged?

seizadi commented:
No, I tried to get this on Kubernetes cluster and found that I needed
logging to work for me to do proper detection of vulnerabilities. I had
used superviord and it made getting native logs out of the container
difficult, so I had to rethink my integration in Kubernetes and have opted
for a more common pattern of breaking this up into its constituents
containers and charts rather than a monolith container. I will let you know
when the project settles down and can be reviewed and merged.

On Wed, Oct 14, 2020 at 1:51 PM Robin Wood <notifications@github.com> wrote:

> Is this ready to be merged?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/digininja/DVWA/pull/384#issuecomment-708653028>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMJXI7F4AJO6ERKRSVVD4LSKYFNTANCNFSM4RQSXE2A>
> .
>

digininja commented:
As far as I know, most people who use DVWA do it at home as a quick to set
up lab. Do you think many people will use this once you get it working?

On Thu, 15 Oct 2020, 00:29 Soheil Eizadi, <notifications@github.com> wrote:

> No, I tried to get this on Kubernetes cluster and found that I needed
> logging to work for me to do proper detection of vulnerabilities. I had
> used superviord and it made getting native logs out of the container
> difficult, so I had to rethink my integration in Kubernetes and have opted
> for a more common pattern of breaking this up into its constituents
> containers and charts rather than a monolith container. I will let you know
> when the project settles down and can be reviewed and merged.
>
> On Wed, Oct 14, 2020 at 1:51 PM Robin Wood <notifications@github.com>
> wrote:
>
> > Is this ready to be merged?
> >
> > —
> > You are receiving this because you modified the open/close state.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/digininja/DVWA/pull/384#issuecomment-708653028>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAMJXI7F4AJO6ERKRSVVD4LSKYFNTANCNFSM4RQSXE2A
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/digininja/DVWA/pull/384#issuecomment-708713487>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4SWMVVMQFOMA3QIXQQD3SKYX4VANCNFSM4RQSXE2A>
> .
>

fldu commented:
@digininja I think it could be useful to have that kind of environment if you want to setup and tests WAFs or test payloads

digininja commented:
OK, that makes sense

On Thu, 15 Oct 2020 at 08:13, Florian Dubourg <notifications@github.com>
wrote:

> @digininja <https://github.com/digininja> I think it could be useful to
> have that kind of environment if you want to setup and tests WAFs or test
> payloads
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/digininja/DVWA/pull/384#issuecomment-708950341>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4SWK33WVWIUQI7DK45DTSK2OIDANCNFSM4RQSXE2A>
> .
>

seizadi commented:
That use case of quick setup will still work, it only takes a single
command line to run the environment. on kubernetes using helm. Kubernetes
is so popular that now docker desktop supports kubernetes:
https://docs.docker.com/docker-for-windows/kubernetes/

You can still run DVWA as a single container, I have just removed MySQL as
bundled with it. If you were to run it using just docker you would have to
run two docker commands one for MySQL and another for the DVWA container.



On Wed, Oct 14, 2020 at 11:42 PM Robin Wood <notifications@github.com>
wrote:

> As far as I know, most people who use DVWA do it at home as a quick to set
> up lab. Do you think many people will use this once you get it working?
>
> On Thu, 15 Oct 2020, 00:29 Soheil Eizadi, <notifications@github.com>
> wrote:
>
> > No, I tried to get this on Kubernetes cluster and found that I needed
> > logging to work for me to do proper detection of vulnerabilities. I had
> > used superviord and it made getting native logs out of the container
> > difficult, so I had to rethink my integration in Kubernetes and have
> opted
> > for a more common pattern of breaking this up into its constituents
> > containers and charts rather than a monolith container. I will let you
> know
> > when the project settles down and can be reviewed and merged.
> >
> > On Wed, Oct 14, 2020 at 1:51 PM Robin Wood <notifications@github.com>
> > wrote:
> >
> > > Is this ready to be merged?
> > >
> > > —
> > > You are receiving this because you modified the open/close state.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/digininja/DVWA/pull/384#issuecomment-708653028>,
> or
> > > unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AAMJXI7F4AJO6ERKRSVVD4LSKYFNTANCNFSM4RQSXE2A
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/digininja/DVWA/pull/384#issuecomment-708713487>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAA4SWMVVMQFOMA3QIXQQD3SKYX4VANCNFSM4RQSXE2A
> >
> > .
> >
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/digininja/DVWA/pull/384#issuecomment-708935518>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMJXI7CIMIP74HCWNFVCFTSK2KVBANCNFSM4RQSXE2A>
> .
>

digininja commented:
I'm going to close this as a lot has changed since this was created and I don't know if it is still valid. If it is, and you still want it, feel free to patch it up and we can try again.
